pandas/core/arrays/timedeltas.py

```diff
@@ -134,55 +133,39 @@ def dtype(self):
     _attributes = ["freq"]
 
     def __init__(self, values, dtype=_TD_DTYPE, freq=None, copy=False):
-        if isinstance(values, (ABCSeries, ABCIndexClass)):
-            values = values._values
-
-        if isinstance(values, type(self)):
-            values, freq, freq_infer = extract_values_freq(values, freq)
-
-        if not isinstance(values, np.ndarray):
-            msg = (
+        if not hasattr(values, "dtype"):
+            raise ValueError(
                 "Unexpected type '{}'. 'values' must be a TimedeltaArray "
                 "ndarray, or Series or Index containing one of those."
-            )
-            raise ValueError(msg.format(type(values).__name__))
-
-        if values.dtype == 'i8':
-            # for compat with datetime/timedelta/period shared methods,
-            #  we can sometimes get here with int64 values.  These represent
-            #  nanosecond UTC (or tz-naive) unix timestamps
-            values = values.view(_TD_DTYPE)
-
-        if values.dtype != _TD_DTYPE:
-            raise TypeError(_BAD_DTYPE.format(dtype=values.dtype))
-
-        try:
-            dtype_mismatch = dtype != _TD_DTYPE
-        except TypeError:
-            raise TypeError(_BAD_DTYPE.format(dtype=dtype))
-        else:
-            if dtype_mismatch:
-                raise TypeError(_BAD_DTYPE.format(dtype=dtype))
-
+                .format(type(values).__name__))
         if freq == "infer":
-            msg = (
+            raise ValueError(
                 "Frequency inference not allowed in TimedeltaArray.__init__. "
-                "Use 'pd.array()' instead."
-            )
-            raise ValueError(msg)
+                "Use 'pd.array()' instead.")
 
-        if copy:
-            values = values.copy()
-        if freq:
-            freq = to_offset(freq)
+        if dtype is not None and not is_dtype_equal(dtype, _TD_DTYPE):
+            raise TypeError("dtype {dtype} cannot be converted to "
+                            "timedelta64[ns]".format(dtype=dtype))
+
+        if values.dtype == 'i8':
+            values = values.view('timedelta64[ns]')
 
-        self._data = values
-        self._dtype = dtype
-        self._freq = freq
+        result = type(self)._from_sequence(values, dtype=dtype,
+                                           copy=copy, freq=freq)
+        self._data = result._data
+        self._freq = result._freq
+        self._dtype = result._dtype
 
     @classmethod
     def _simple_new(cls, values, freq=None, dtype=_TD_DTYPE):
-        return cls(values, dtype=dtype, freq=freq)
+        assert dtype is _TD_DTYPE, dtype
```

Comment:
`np.dtype('m8[ns]')` isn't a singleton, so  this may fail on `dtype=np.dtype("m8[ns]")` (or even for `dtype="m8[ns]")`. Do we want to prohibit passing dtype entirely?